### ISSUE TITLE
Having some issues with embed tag

### ISSUE THREAD

**Issue by lvl99** [twigjs/twig.js#352]
I'm doing a bunch of extending, including and embedding, and I'm finding when I use `embed` tags that it seems to ignore the content within them and further affects other block content within the same layout.

My structure is as follows:

```
  **layout.twig:**
    block header
      include partials/header.twig

    block content

    block footer
      include partials/footer.twig
```

I have another layout template which extends `layout.twig` and adds some additional blocks, as well as overwriting the `layout.twig`'s main header, content and footer blocks. In this layout, I'm using embeds in order to change specific areas within the header and footer:

```
  **dashboard.layout.twig extends layout.twig:**
    block header
      embed partials/header.twig
         (*contains block overrides for specific blocks within header.twig*)

    block content
       block dashboardContent

    block footer
      embed partials/footer.twig
        (*contains block overrides for specific blocks within footer.twig*)
```

Since the above is a reusable layout file, it's not one that I use for a specific page to be rendered to HTML. Here is the actual page's structure for rendering:

```
  **page.dashboard.layout.twig extends dashboard.layout.twig:**
    block dashboardContent
      (*custom content for this specific page*)
```

When the `page.dashboard.layout.twig` file is rendered, the header and the footer are unmodified (essentially same as default `layout.twig`) ignoring the embed block overrides, although it has rendered the `dashboard.twig` layout correctly, albeit with none of the content defined in the `dashboardContent` block.

When I remove the header/footer blocks or the `embed` tags from the `dashboard.layout.twig` structure, the `page.dashboard.layout.twig` block `dashboardContent` is correctly rendered for the page. Is it because I'm using two levels of extends that this might be causing some kind of issue? Are extends only designed for a single level of inheritance?

**Comment by dave-irvine** [comment]
If it works in PHP twig, it should be working here. Does your example work with PHP twig? If so, could you try and simplify down to the simplest breaking case and reply with an example and we will try and figure it out.

**Comment by lvl99** [comment]
I've made a quick version of the example I've given above. It seems to render correctly with all modifications cascading correctly: http://twigfiddle.com/vk11o4

I'm just going to run this example through Twig.JS now. Will report back soon.

**Comment by lvl99** [comment]
It looks like Twig.JS doesn't seem to cascade correctly with the embeds:

![screen shot 2016-05-03 at 12 57 46 pm](https://cloud.githubusercontent.com/assets/238480/14981338/c70661e0-112e-11e6-9c60-2ba5f6d15d00.png)

**Comment by dave-irvine** [comment]
Alrighty, bug then!

I'll try and track this one down.

**Comment by lvl99** [comment]
Much appreciated!

**Comment by lvl99** [comment]
I've just been running some debug on the above example. I've done one with the `embed` tags present, and one with `include` tags. Side note: in my personal file notation, files starting with an underscore generally denote extendable layout templates, not actual rendered-to-HTML pages.

[twig-embeds-debug-example-embed.txt](https://github.com/twigjs/twig.js/files/246919/twig-embeds-debug-example-embed.txt)
[twig-embeds-debug-example-include.txt](https://github.com/twigjs/twig.js/files/246918/twig-embeds-debug-example-include.txt)

**Comment by dave-irvine** [comment]
Ok with my simplest possible example:

### BugSnip|index.js
```javascript
var Twig = require('twig'), // Twig module
    twig = Twig.twig;       // Render function

twig({
	id:   'extends',
	path: 'main.template.twig',
	async: false
});

var template = twig({ref: 'extends'})
console.log(template.render());
```
###BugSnip|main.template.twig
```
{% extends 'sub.template.twig' %}
```
###BugSnip|sub.template.twig
```
{% block header %}
    {% embed 'template.twig' %}
    {% block headerTitle %}Dashboard{% endblock %}
    {% endembed %}
{% endblock %}
```
###BugSnip|template.twig
```
<header>
    <h1>{% block headerTitle %}Layout title{% endblock %}</h1>
</header>
```

###### BugSnip Build Status (atilt-dad)
[![BugSnippet](https://www.bugsnip.com/images/atilt-dad.gif)](https://www.bugsnip.com/logs/atilt-dad.log)

**Comment by dave-irvine** [comment]
###### BugSnip Build Status (museful-house)

**Comment by lvl99** [comment]
Your example works, but it's not quite the same use-case as I'm doing two levels of `extend`ing. Adapting your example:

**main.template.twig**
```
{% extends 'sub.template.twig' %}

{% block mainContent %}
  <p>Main: content.mainContent</p>
{% endblock %}
```

**sub.template.twig**
```
{% extends 'base.template.twig' %}

{% block header %}
  {% embed 'header.twig' %}
    {% block headerTitle %}Sub: header.headerTitle{% endblock %}
  {% endembed %}
{% endblock %}

{% block content %}
  {% block mainContent %}
    <p>Sub: content.mainContent</p>
  {% endblock %}
{% endblock %}
```

**base.template.twig**
```
{% block header %}
  {% include 'header.twig' %}
{% endblock %}

{% block content %}
  <p>Base: content</p>
{% endblock %}
```

**header.twig**
```
<header><h1>{% block headerTitle %}Header: headerTitle{% endblock %}</h1></header>
```

The expected output should be ([View this example on twigfiddle](http://twigfiddle.com/j68lxz)):
```
<header><h1>Sub: header.headerTitle</h1></header>
    <p>Main: content.mainContent</p>
```

The output I get from Twig.JS is:
```
  <header><h1>Sub: header.headerTitle</h1></header>
      <p>Sub: content.mainContent</p>
```

Perhaps it's just not processing the `extend`s and various block overrides in the right order?

**Comment by dave-irvine** [comment]
If I remove

```
{% block header %}
  {% embed 'header.twig' %}
    {% block headerTitle %}Sub: header.headerTitle{% endblock %}
  {% endembed %}
{% endblock %}
```

from sub.template.twig, then I get the following output:

```
  <header><h1>Header: headerTitle</h1></header>
    <p>Main: content.mainContent</p>
```

which is pretty much what I would expect. So this leaves me with the question of if it is `embed` breaking things.

**Comment by dave-irvine** [comment]
Ok if I change `base.template.twig` to be:

##### base.template.twig
```
{% block header %}
  {% embed 'header.twig' %}
    {% block headerTitle %}Base: header.headerTitle{% endblock %}
  {% endembed %}
{% endblock %}

{% block content %}
  <p>Base: content</p>
{% endblock %}
```

I get:

```
  <header><h1>Base: header.headerTitle</h1></header>

  <p>Base: content</p>
```

So I'm pretty sure its the embed breaking stuff. Which means some deep-dive debugging. My fave :)

**Comment by dave-irvine** [comment]
@lvl99 Can you see if #369 has magically fixed your issue? I was aiming to fix something else but it might have the side effect of fixing this too!

**Comment by lvl99** [comment]
Unfortunately it doesn't! I also went from a 4~8sec compilation time (admittedly using `gulp-twig` which is using `twig@0.8.6`) to ~40secs!

**Comment by dave-irvine** [comment]
Is it actually using 0.8.6 or has it updated itself to 0.8.9?

My latest changes landed in 0.9.5 so I was hoping you could try that out.

**Comment by lvl99** [comment]
It was using 0.8.6 and I manually updated it to use 0.9.5

**Comment by dave-irvine** [comment]
Can you join us on gitter? https://gitter.im/twigjs/twig.js

**Comment by name-k** [comment]
Guys, did you find any solution? I have the same thing going on. 

My page structure is 
``` twig
{% extends 'layout.twig' %}

{% block options %}
  {% set var = 'value' %}
{% endblock %}

{% embed "embed.twig" %}
```

In layout.twig 
``` twig
{{ var == null }}
```

When there is an embed, "true" will be printed, without embed - "false", which is expected to be.

**Comment by olets** [comment]
@name-k short answer is no, the embed tag hasn't been fixed.

But could you give us some more context? What is embed.twig? And it looks like you're missing `{% endembed %}`?

**Comment by name-k** [comment]
@olets sure, there is ```{% endembed %}```, and embed.twig is an empty file, I found that its content does not matter. 

My workaround is to place all blocks in the page after all the embed, in my case it looks like that
```
{% extends 'layout.twig' %}

{% embed "embed.twig" %}{% endembed %}

{% block options %}
  {% set var = 'value' %}
{% endblock %}
```
And everything works fine now. Though sometimes it is not an option, when order is important.

**Comment by Jones-S** [comment]
As I understand, the embed-issue is not fixed yet. Does anyone have any idea how to tackle this problem? We need this to work in one of our project and we may want to work on a solution... So any starting point is very welcome.

**Comment by aclaes** [comment]
We ran into the same issue with nested embeds messing around. 
It seems like there is currently not a lot of attention on this topic. 
However, if someone is wrapping his head around this, I created another simple example to demonstrate the error.  Maybe it helps to tackle down the issue.

Example:
https://twigfiddle.com/rjatuc

Expected result (PHP implementation):
``` html
<div class="outer">
    <div class="inner">
        foo
    </div>
    <div class="inner">
        bar
    </div>
</div>
```

Actual result (JS implementation):
``` html
<div class="outer">
    <div class="inner">
        bar
    </div>
</div>
```

As you can see, the second usage of `{% embed "inner.twig" %}` kills the first on. 
If you would have multiple inner embeds within the outer embed, only the last one will make ie to the output.

**Comment by aclaes** [comment]
I've been reading and debugging the JS implementation a bit. I think, I got close to the issue, but I don't know where to go next.

As I understood, the way `embed` breaks things, is by deleting `blocks`.  This is done very explicitly:
https://github.com/twigjs/twig.js/blob/master/src/twig.logic.js#L1108

At this point I don't find a reason for this behaviour. 

However, If a simply remove this line and run my minimal example above, I encounter a second problem:

Within the inner `embed` the `block` named "content" doesn't get overwritten. It renders two times to "foo" instead of "foo", then "bar"

Maybe this is helpful in any way...

**Comment by JuSenpai** [comment]
Anything new about this? I've encountered this bug and have to find workarounds because of it... Will it be fixed soon?

**Comment by toptalo** [comment]
Hello! Same issue

**Comment by toptalo** [comment]
Found this: https://github.com/twigjs/twig.js/blob/master/src/twig.logic.js#L1110
```javascript
	// reset previous blocks
	that.blocks = {};
```

If I comment this line -> all works. But need someone who know reason of this line

**Comment by toptalo** [comment]
This also helps
```javascript
    // reset previous blocks
    that._blocks = Object.assign({}, that.blocks);
    that.blocks = {};
    
    // parse tokens. output will be not used
    return Twig.parseAsync.apply(that, [token.output, innerContext])
        .then(function() {
            that.blocks = Object.assign(that._blocks, that.blocks);
            // render tempalte with blocks defined in embed block
            return template.renderAsync(innerContext, {'blocks':that.blocks});
        });
```

**Comment by toptalo** [comment]
@dave-irvine @lvl99 Can you check my [PR](#537)

**Comment by toptalo** [comment]
I think, this issue is the same as #287

**Comment by willrowe** [comment]
- I believe you are correct @toptalo I have a test case for this as well as the one you provided in #287 included in an upcoming PR.
- I will have the mentioned PR ready to go this week hopefully, it should pave the way for handling blocks more cleanly.

**Comment by willrowe** [comment]
This was fixed by #550.
